addons/account/models/account_move.py

```diff
@@ -1566,19 +1569,25 @@ def _fetch_duplicate_supplier_reference(self, matching_states=('draft', 'posted'
               FROM {move_table_and_alias}
               JOIN account_move AS duplicate_move ON
                    move.company_id = duplicate_move.company_id
+               AND move.id != duplicate_move.id
+               AND duplicate_move.state IN %(matching_states)s
+               AND move.move_type = duplicate_move.move_type
                AND (
                    move.commercial_partner_id = duplicate_move.commercial_partner_id
                    OR (move.commercial_partner_id IS NULL AND duplicate_move.state = 'draft')
-               )
-               AND move.ref = duplicate_move.ref
-               AND move.move_type = duplicate_move.move_type
-               AND move.id != duplicate_move.id
-               AND (move.invoice_date = duplicate_move.invoice_date OR move.state = 'draft')
-               AND duplicate_move.state IN %(matching_states)s
+                )
+               AND (
+                   -- For out moves
+                   (move.move_type in ('out_invoice', 'out_refund')
+                   AND (move.amount_total = duplicate_move.amount_total AND move.invoice_date = duplicate_move.invoice_date))
+                    OR 
+                   -- For in moves
+                   (move.move_type in ('in_invoice', 'in_refund') AND (move.ref = duplicate_move.ref AND (move.invoice_date = duplicate_move.invoice_date OR move.state = 'draft')))
+               ) 
```

Comment:
What do you think about formatting it like this?
```suggestion
               AND (
                   move.move_type in ('out_invoice', 'out_refund')
                   AND (
                       move.amount_total = duplicate_move.amount_total
                       AND move.invoice_date = duplicate_move.invoice_date
                   )
                   OR
                   move.move_type in ('in_invoice', 'in_refund')
                   AND (
                       move.ref = duplicate_move.ref
                       AND (move.invoice_date = duplicate_move.invoice_date OR move.state = 'draft')
                   )
               )
```